File: BE/src/main/java/it/polimi/iswpf/service/_interface/RecensioneService.java
package it.polimi.iswpf.service._interface;

import it.polimi.iswpf.dto.request.InviaRecensioneRequest;
import it.polimi.iswpf.dto.response.RecensioneDettagliataResponse;
import it.polimi.iswpf.dto.response.RecensioneResponse;
import it.polimi.iswpf.service.implementation.RecensioneServiceImpl;

import java.util.List;

/**
 * Interfaccia che contiene le firme dei metodi del service che si occupa delle recensioni.
 * Implementazione -> {@link RecensioneServiceImpl}.
 */
public interface RecensioneService {

    /**
     * Metodo per salvare sul database una recensione di un utente.
     * @param request DTO con i dati della recensione.
     * @param eventoId Id dell'evento, passato in modo dinamico tramite l'endpoint.
     * @param turistaId Id del turista, passato in modo dinamico tramite l'endpoint.
     */
    void inviaRecensione(InviaRecensioneRequest request, Long eventoId, Long turistaId);

    /**
     * Metodo per prendere tutte le recensioni di un dato evento.
     * @param eventoId Id univoco dell'evento, passato in modo dinamico tramite l'endpoint.
     * @return Lista di DTO con i dati delle recensioni.
     */
    List<RecensioneResponse> getByEvento(Long eventoId);

    /**
     * Metodo per prendere tutti i dati di una singola recensione.
     * @param eventoId Id univoco dell'evento, passato in modo dinamico tramite l'endpoint.
     * @param usernameTurista Username univoco del turista che ha recensito l'evento,
     * passato in modo dinamico tramite l'endpoint.
     * @return DTO con i dati della recensione.
     */
    RecensioneDettagliataResponse getRecensione(Long eventoId, String usernameTurista);
}

File: FE/src/app/services/recensione.service.ts
import { HttpClient, HttpHeaders } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { inviaRecensioneRequest } from '../dtos/request/InviaRecensioneRequest';
import { RecensioneResponse } from '../dtos/response/RecensioneResponse';
import { RecensioneDettagliataResponse } from '../dtos/response/RecensioneDettagliataResponse';
import { MessageResponse } from '../dtos/response/MessageResponse';
import { globalBackendUrl } from 'environment';

/**
 * service per gestire le interazioni col backend riguardo le recensioni
 */
@Injectable({
  providedIn: 'root'
})
export class RecensioneService {
  private backendUrl: string = globalBackendUrl + 'recensione/';

  //costruttore dove istanzio le classi con cui interagire
  constructor(private http: HttpClient) { }

  /**
   * metodo per salvare una recensione sul database
   * @param eventoId id dell'evento
   * @param voto voto della recensione
   * @param testo testo della recensione
   * @returns messaggio di avvenuto invio della recensione
   */
  inviaRecensione(eventoId: string, voto: number, testo: string): Observable<MessageResponse> {
    const header = this.getHeader();
    const request: inviaRecensioneRequest = { voto, testo };

    return this.http.post<MessageResponse>(this.backendUrl + 'inviaRecensione/' + eventoId + '/' + localStorage.getItem('id')?.toString().trim(), request, { headers: header });
  }

  /**
   * chiamo il backend per prendere le recensioni di un dato evento
   * @param eventoId id dell'evento
   * @returns array di DTO con le recensioni
   */
  getRecensioniByEvento(eventoId: string): Observable<RecensioneResponse[]> {
    const header = this.getHeader();

    return this.http.get<RecensioneResponse[]>(this.backendUrl + 'getByEvento/' + eventoId, { headers: header });
  }

  /**
   * chiamo il backend per prendere i dati dettagliati di una singola recensione
   * @param eventoId id dell'evento
   * @param usernameTurista useranme del turista
   * @returns DTO con i dati dettagliati della singola recensione
   */
  getRecensione(eventoId: string, usernameTurista: string): Observable<RecensioneDettagliataResponse> {
    const header = this.getHeader();

    return this.http.get<RecensioneDettagliataResponse>(this.backendUrl + 'get/' + eventoId + '/' + usernameTurista, { headers: header });
  }

  //creo l'header con il token da mandare al backend
  private getHeader(): HttpHeaders {
    return new HttpHeaders({
      'Authorization': localStorage.getItem('token') ? `${localStorage.getItem('token')}` : '',
      id: localStorage.getItem('id') ? `${localStorage.getItem('id')}` : '',
      ruolo: localStorage.getItem('ruolo') ? `${localStorage.getItem('ruolo')}` : ''
    });
  }
}
